source/particle/world.cc

```diff
@@ -206,6 +210,88 @@ namespace aspect
+        {
+
+          // Loop over all cells and update the particles cell-wise
+          typename DoFHandler<dim>::active_cell_iterator
+          cell = this->get_dof_handler().begin_active(),
+          endc = this->get_dof_handler().end();
+
+          for (; cell!=endc; ++cell)
+            if (cell->is_locally_owned())
+              {
+                const types::LevelInd found_cell(cell->level(),cell->index());
+
+                const unsigned int particles_in_cell = particles.count(found_cell);
+                // Add particles if necessary
+                if (particle_load_balancing == remove_and_add_particles)
+                  {
+                    for (unsigned int i = particles_in_cell; i < min_particles_per_cell; ++i)
+                      {
+                        std::pair<aspect::Particle::types::LevelInd,Particle<dim> > new_particle = generator->generate_particle(cell,0);
+
+                        Vector<double> value(this->introspection().n_components);
+                        std::vector<Tensor<1,dim> > gradient (this->introspection().n_components,Tensor<1,dim>());
+
+                        std::vector<Vector<double> >  solution(1,value);
+                        std::vector<std::vector<Tensor<1,dim> > > gradients(1,gradient);
+
+                        std::vector<Point<dim> >     particle_points(1);
+                        particle_points[0] = this->get_mapping().transform_real_to_unit_cell(cell, new_particle.second.get_location());
+
+                        const Quadrature<dim> quadrature_formula(particle_points);
+                        FEValues<dim> fe_value (this->get_mapping(),
+                                                this->get_fe(),
+                                                quadrature_formula,
+                                                update_values |
+                                                update_gradients);
+
+                        fe_value.reinit (cell);
+                        fe_value.get_function_values (this->get_solution(),
+                                                      solution);
+                        fe_value.get_function_gradients (this->get_solution(),
+                                                         gradients);
+
+                        property_manager->initialize_late_particle(new_particle.second,
+                                                                   particles,
+                                                                   *interpolator,
+                                                                   solution[0],
+                                                                   gradients[0]);
+
+                        particles.insert(new_particle);
+                      }
+                  }
+
+                // Remove particles if necessary
+                if ((max_particles_per_cell > 0) && (particles_in_cell > max_particles_per_cell))
```

Comment:
this case is mutually exclusive to the above case -- maybe insert an `else` before the `if`?